core/src/main/java/io/grpc/Context.java

```diff
@@ -434,13 +422,16 @@ public void addListener(final CancellationListener cancellationListener,
    * Remove a {@link CancellationListener}.
    */
   public void removeListener(CancellationListener cancellationListener) {
+    if (!canBeCancelled()) {
```

Comment:
Given that `canBeCancelled()` walks the tree, I'm thinking this may not be an optimization. Acquiring an uncontested lock may be faster.